files/usr/share/jeos-firstboot/modules/ssh_enroll

```diff
@@ -0,0 +1,131 @@
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: Copyright 2024 SUSE LLC
+# shellcheck shell=bash
+
+ssh_enroll_handle_pubkeys()
+{
+	filename="$1"
+	if ! [ -s "$filename" ]; then
+		d --msgbox $"No public keys received!" 0 0
+		return 0
+	fi
+
+	items=()
+	lineno=1
+	state=on
+	while read -r type key comment; do
+		items+=("$lineno" "${type} ${key:0:8}...${key: -20:20}" "$state")
+		lineno=$((lineno+1))
+		state=off # Only import the first key by default
+	done <"$filename"
+
+	text="$(printf $"Received %d public keys from %s. Select the key(s) to import." "$(wc -l <"$filename")" "$(awk '{print $3; exit}' "$filename")")"
+	d --title $"SSH Key Enrollment" --separate-output --no-tags --checklist "$text" 12 60 10 "${items[@]}"
+
+	readarray -t input <<<"$result"
+	# readarray converts an empty string to ("")
+	if [ -z "${input[0]}" ]; then
+		return 0
+	fi
+
+	(umask 077; mkdir -p ~/.ssh)
+	for lineno in "${input[@]}"; do
+		awk "NR==${lineno}" <"$filename" >> ~/.ssh/authorized_keys
+	done
+
+	text="$(printf $"Imported %d keys" "${#input[@]}")"
+	d --title $"SSH Key Enrollment" --msgbox "$text" 0 0
+}
+
+ssh_enroll_do_config()
+{
+	if ! dialog --backtitle "$PRETTY_NAME" --title $"SSH Key Enrollment" --yesno $"Do you want to enroll keys for SSH access?" 0 0; then
+		return 0
+	fi
+
+	# Stop sshd.service temporarily if needed
+	start_sshd_again=0
+	if systemctl -q is-active sshd.service; then
+		systemctl stop sshd.service
+		start_sshd_again=1
+	fi
+
+	# Make sure host keys exist
+	sshd-gen-keys-start &>/dev/null
+
+	tmpdir=$(umask 077; mktemp -d)
+
+	# Start the server in the background and show a dialog with information while it's running
+	# and offer the option to cancel the process. Unfortunately bash doesn't have a way to wait for
+	# one of multiple processes to exit and get information about which process exited and how.
+	# To work around that, use subprocesses with FIFOs: One to signal once they're up and can handle
+	# SIGTERM properly and one to signal that it exited and how.
+	# See also the welcome-screen implementation.
+	mkfifo "${tmpdir}/serverfifo"
+	mkfifo "${tmpdir}/dialogfifo"
+	mkfifo "${tmpdir}/exitfifo"
```

Comment:
Actually I believe this *is* possible in bash like this:

```
sleep $(( $RANDOM % 20 ))  &
SLEEP1_PID=$!
sleep $(( $RANDOM % 20 )) &
SLEEP2_PID=$!

wait -n -p which %1 %2
res=$?

if [ "$which" -eq "$SLEEP1_PID" ]; then
        echo "sleep1 ($which) exited with $res"
else
        echo "sleep2 ($which) exited with $res"
fi

wait
```

Without the FIFOs and sub shells the code will become simpler.

Reply:
Nice, I only had a look at the POSIX/sh wait documentation. This might be enough to work for this case.

In another branch (`ssh_enroll_simple`) I tried an even simpler method which requires the user to use Ctrl-C to abort the process, which is the one I'm currently favoring. I'll try this out in any case.

Reply:
> Nice, I only had a look at the POSIX/sh wait documentation. This might be enough to work for this case.

Yep, done.

> In another branch (`ssh_enroll_simple`) I tried an even simpler method which requires the user to use Ctrl-C to abort the process, which is the one I'm currently favoring. I'll try this out in any case.

The inclusion of randomart means that scrolling might be necessary, so the simpler method no longer works with that.